Refactor/net: split headers to proxy class

Ryex opened the pull request:
This PR splits the setting of headers out of the Net::Download class into a proxy class. This allows a more generic controll flow to use the download class without invoking a static cast on the QCoreApplication::inst() pointer 

(I'm trying to let things like the updater use the networks tasks)

Ryex committed:
refactor: split out setting api headers for downloads

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
refactor: override / mask static `make` functions for ApiDownload

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
chore: format

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix: class rename (lsp acitons undid the first rename :P)

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

getchoo commented:
is this complete? just wondering since i noticed the old `Net::Download` functions weren't removed yet, and i'm not sure if that's for backwards compatibility. otherwise lgtm :)

Ryex committed:
feat(net): ApiUpload ^& fix unfired `finished` signals

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
Merge remote-tracking branch 'upstream/develop' into refactor/net-split-headers-to-proxy-class

Ryex commented:
> is this complete? just wondering since i noticed the old `Net::Download` functions weren't removed yet, and i'm not sure if that's for backwards compatibility. otherwise lgtm :)

> is this complete? just wondering since i noticed the old `Net::Download` functions weren't removed yet, and i'm not sure if that's for backwards compatibility. otherwise lgtm :)

Ok this is fixed and finished.
Needed to add a compiler def to deal with the user agent and apply the changed to the `Upload` class too.

Ryex committed:
fix: add `ApiUpload` to CMakeLists.txt

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fixup! Merge remote-tracking branch 'upstream/develop' into refactor/net-split-headers-to-proxy-class

Ryex committed:
DCO Remediation Commit for Rachel Powers <508861+Ryex@users.noreply.github.com>

I, Rachel Powers <508861+Ryex@users.noreply.github.com>, hereby add my Signed-off-by to this commit: 1ba08f4641f32d39efc509247807eca035be62f5

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

getchoo committed:
fix(net): fix emit signals in download & upload task

Signed-off-by: seth <getchoo@tuta.io>

Ryex committed:
fix: ensure `finished` signal is emited

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
> is this complete? just wondering since i noticed the old `Net::Download` functions weren't removed yet, and i'm not sure if that's for backwards compatibility. otherwise lgtm :)

the old functions are still used in some places and outside the the main launcher in the updater I'm building. which is the whole point of making this change.

Trial97 committed:
Simplify Upload and Download

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented:
Maybe you can also unify Upload and Download as mostly the code base is the same.
Here is my idea: https://github.com/Ryex/PrismLauncher/compare/refactor/net-split-headers-to-proxy-class...Trial97:PrismLauncher:refactor/net-split-headers-to-proxy-class?expand=1

Ryex commented:
> Maybe you can also unify Upload and Download as mostly the code base is the same. Here is my idea: https://github.com/Ryex/PrismLauncher/compare/refactor/net-split-headers-to-proxy-class...Trial97:PrismLauncher:refactor/net-split-headers-to-proxy-class?expand=1

Id actually very much like that refactor. can you clean up that patch though? there seems to be a bunch of extraneous stuff in it.

Trial97 committed:
More clenup

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented:
> Id actually very much like that refactor. can you clean up that patch though? there seems to be a bunch of extraneous stuff in it.

Done https://github.com/Ryex/PrismLauncher/pull/1

Ryex committed:
Merge pull request #1 from Trial97/refactor/net-split-headers-to-proxy-class

Simplify Upload and Download

Ryex commented:
> > Id actually very much like that refactor. can you clean up that patch though? there seems to be a bunch of extraneous stuff in it.
> 
> Done [Ryex#1](https://github.com/Ryex/PrismLauncher/pull/1)

ty

Trial97 committed:
Fail NetRequest on connection close from QT

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

getchoo reviewed (approved):
alright, previous issues look to be resolved; after running a few sets of downloads/updates from cf and modrinth, i haven't been able to find any regressions! the refactor also makes a lot of sense imo, along with the new netrequest class (and thanks trial for making that a separate pr, made it really easy to review for me :p)

adding this to the 8.0 milestone since a few others prs are dependent on it now

Ryex committed:
Merge pull request #3 from Trial97/netRefactor3

Fail NetRequest on connection close from QT

Ryex committed:
Merge branch 'develop' into refactor/net-split-headers-to-proxy-class

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex committed:
fix(FlameInstanceCreationTask): include Net::Apidownload

Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

TheKodeToad commented on launcher/net/ApiDownload.cpp:
```diff
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  Prism Launcher - Minecraft Launcher
+ *  Copyright (C) 2023 Rachel Powers <508861+Ryex@users.noreply.github.com>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, version 3.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
```
You should probably include MultiMC's license, as well as in other places (along with any applicable contributors)

Ryex replied:
Nah, the code in this fine is original. `NetRequest` now holds the reused code. which is why it has the referenced license.

TheKodeToad replied:
The `make*` functions look very similar to in Download.cpp

Ryex replied:
similar in name only. conforming to the same interface doesn't mean they aren't original.

TheKodeToad replied:
huh? the function implementations look almost the same

Ryex replied:
There are only so many ways to write them. They are new functions on a new class, the old functions still exist, there for these are original code. This is actually law in the US thanks to Oracle vs Google.

TheKodeToad commented on launcher/net/ApiHeaderProxy.h:
```diff
@@ -0,0 +1,49 @@
+        QList<HeaderPair> hdrs;
+        if (APPLICATION->capabilities() & Application::SupportsFlame && request.url().host() == QUrl(BuildConfig.FLAME_BASE_URL).host()) {
+            hdrs.append({ "x-api-key", APPLICATION->getFlameAPIKey().toUtf8() });
+        } else if (request.url().host() == QUrl(BuildConfig.MODRINTH_PROD_URL).host() ||
+                   request.url().host() == QUrl(BuildConfig.MODRINTH_STAGING_URL).host()) {
```
(note) Should be moved in future

TheKodeToad commented on launcher/net/ApiHeaderProxy.h:
```diff
@@ -0,0 +1,49 @@
+
+class ApiHeaderProxy : public HeaderProxy {
+   public:
+    ApiHeaderProxy() : HeaderProxy(){};
```
```suggestion
    ApiHeaderProxy() : HeaderProxy() {}
```

TheKodeToad commented on launcher/net/HeaderProxy.h:
```diff
@@ -0,0 +1,49 @@
+        for (auto header : headers(request)) {
+            request.setRawHeader(header.headerName, header.headerValue);
+        }
+    };
```
```suggestion
    }
```

TheKodeToad commented on launcher/net/NetRequest.h:
```diff
@@ -0,0 +1,99 @@
+   public:
+    void addValidator(Validator* v);
+    auto abort() -> bool override;
+    auto canAbort() const -> bool override { return true; };
```
```suggestion
    auto canAbort() const -> bool override { return true; }
```
\>:(

TheKodeToad commented on launcher/net/RawHeaderProxy.h:
```diff
@@ -0,0 +1,44 @@
+
+class RawHeaderProxy : public HeaderProxy {
+   public:
+    RawHeaderProxy() : HeaderProxy(){};
```
```suggestion
    RawHeaderProxy() : HeaderProxy() {}
```

TheKodeToad commented on launcher/net/Download.h:
```diff
@@ -38,57 +38,26 @@
-
-   public:
-    ~Download() override = default;
+    explicit Download() : NetRequest() { logCat = taskDownloadLogC; };
```
```suggestion
    explicit Download() : NetRequest() { logCat = taskDownloadLogC; }
```

Ryex committed:
Apply suggestions from code review

Co-authored-by: TheKodeToad <TheKodeToad@proton.me>
Signed-off-by: Rachel Powers <508861+Ryex@users.noreply.github.com>

Ryex commented:
ok, forgiven.

Scrumplex commented:
I don't trust my merge commits tbh. I used JetBrains' magic merge tool, and it might have done some questionable stuff. Let's just move it to staging instead
